File: backend/fastapi/app/core/http_client.py
"""
HTTP客户端管理核心模块
"""

from typing import Dict, Optional
import httpx
from fastapi import HTTPException

from app.config import settings


class HTTPClientManager:
    """HTTP客户端管理器"""
    
    def __init__(self):
        self.client: Optional[httpx.AsyncClient] = None
    
    async def init_client(self):
        """初始化HTTP客户端"""
        self.client = httpx.AsyncClient(
            timeout=httpx.Timeout(settings.REQUEST_TIMEOUT),
            limits=httpx.Limits(
                max_keepalive_connections=settings.MAX_KEEPALIVE_CONNECTIONS,
                max_connections=settings.MAX_CONNECTIONS
            )
        )
    
    async def close_client(self):
        """关闭HTTP客户端"""
        if self.client:
            await self.client.aclose()
    
    async def get(self, url: str, params: Dict = None, headers: Dict = None) -> Dict:
        """GET请求"""
        try:
            if not self.client:
                raise HTTPException(status_code=503, detail="HTTP客户端未初始化")
            
            response = await self.client.get(url, params=params or {}, headers=headers or {})
            response.raise_for_status()
            return response.json()
        except httpx.RequestError as e:
            print(f"HTTP请求失败: {e}")
            raise HTTPException(status_code=503, detail="网络请求失败")
        except httpx.HTTPStatusError as e:
            print(f"HTTP状态错误: {e}")
            raise HTTPException(status_code=e.response.status_code, detail="服务响应错误")
    
    async def post(self, url: str, json_data: Dict = None, headers: Dict = None) -> Dict:
        """POST请求"""
        try:
            if not self.client:
                raise HTTPException(status_code=503, detail="HTTP客户端未初始化")
            
            response = await self.client.post(
                url, 
                json=json_data or {}, 
                headers=headers or {"Content-Type": "application/json"}
            )
            response.raise_for_status()
            return response.json()
        except httpx.RequestError as e:
            print(f"HTTP请求失败: {e}")
            raise HTTPException(status_code=503, detail="网络请求失败")
        except httpx.HTTPStatusError as e:
            print(f"HTTP状态错误: {e}")
            raise HTTPException(status_code=e.response.status_code, detail="服务响应错误")
    
    async def health_check(self, url: str) -> bool:
        """健康检查请求"""
        try:
            if not self.client:
                return False
            
            response = await self.client.get(url)
            return response.status_code == 200
        except Exception:
            return False


# 全局HTTP客户端管理器实例
http_client_manager = HTTPClientManager() 

File: backend/hono/src/core/http-client.ts
/**
 * HTTP客户端管理核心模块
 */

import { settings } from '../config';
import { ServiceError } from '../types';

class HTTPClientManager {
  private abortController: AbortController | null = null;

  async init(): Promise<void> {
    // HTTP客户端无需特殊初始化
    console.log('HTTP客户端管理器初始化完成');
  }

  async close(): Promise<void> {
    if (this.abortController) {
      this.abortController.abort();
      this.abortController = null;
    }
  }

  async get<T = any>(url: string, params?: Record<string, any>, headers?: Record<string, string>): Promise<T> {
    try {
      const urlWithParams = new URL(url);
      
      if (params) {
        Object.entries(params).forEach(([key, value]) => {
          urlWithParams.searchParams.append(key, String(value));
        });
      }

      const response = await fetch(urlWithParams.toString(), {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          ...headers,
        },
        signal: AbortSignal.timeout(settings.http.timeout),
      });

      if (!response.ok) {
        const error = new Error(`HTTP错误: ${response.status} ${response.statusText}`) as ServiceError;
        error.statusCode = response.status;
        throw error;
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          const timeoutError = new Error('请求超时') as ServiceError;
          timeoutError.statusCode = 408;
          throw timeoutError;
        }
        
        const serviceError = error as ServiceError;
        serviceError.statusCode = serviceError.statusCode || 503;
        throw serviceError;
      }
      
      const unknownError = new Error('网络请求失败') as ServiceError;
      unknownError.statusCode = 503;
      throw unknownError;
    }
  }

  async post<T = any>(url: string, data?: any, headers?: Record<string, string>): Promise<T> {
    try {
      const requestInit: RequestInit = {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...headers,
        },
        signal: AbortSignal.timeout(settings.http.timeout),
      };
      
      if (data) {
        requestInit.body = JSON.stringify(data);
      }
      
      const response = await fetch(url, requestInit);

      if (!response.ok) {
        const error = new Error(`HTTP错误: ${response.status} ${response.statusText}`) as ServiceError;
        error.statusCode = response.status;
        throw error;
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          const timeoutError = new Error('请求超时') as ServiceError;
          timeoutError.statusCode = 408;
          throw timeoutError;
        }
        
        const serviceError = error as ServiceError;
        serviceError.statusCode = serviceError.statusCode || 503;
        throw serviceError;
      }
      
      const unknownError = new Error('网络请求失败') as ServiceError;
      unknownError.statusCode = 503;
      throw unknownError;
    }
  }

  async healthCheck(url: string): Promise<boolean> {
    try {
      const response = await fetch(url, {
        method: 'GET',
        signal: AbortSignal.timeout(settings.http.timeout),
      });
      return response.ok;
    } catch {
      return false;
    }
  }
}

export const httpClientManager = new HTTPClientManager(); 